### ISSUE TITLE
Command-line window mode to enable faster navigation of completion items

### ISSUE THREAD

**Issue by gnosticscrounger** [The-Compiler/qutebrowser#1463]
In Vim, there is a mode which allows you to navigate the history of : commands through a buffer with access to all the standard normal mode keys.  It would be nice to see this feature for qutebrowser, especially in it's current stage where you cannot specify whether you are search for history or bookmarks.  Control-F is my suggested binding for this mode when in the command line.

**Comment by The-Compiler** [comment]
> In Vim, there is a mode which allows you to navigate the history of : commands through a buffer with access to all the standard normal mode keys.

How do I access that in vim so I can take a look at it?

>  It would be nice to see this feature for qutebrowser, especially in it's current stage where you cannot specify whether you are searching for history or bookmarks.

I don't see how that's related to history/bookmarks - this is only about command history, correct?

**Comment by gnosticscrounger** [comment]
Go into command line mode (via :) and then press Control-F. You should open up a buffer that has the history of commands on a series of lines. 

Sorry for not clarifying this, I would like this both for completion items and command history. This should be context sensitive to whether I have an open command currently unfinished or if it is blank.

**Comment by The-Compiler** [comment]
Thanks for the clarification. This would be quite a lot of effort for (IMHO) small benefit, so I'm not sure about this yet.

**Comment by gnosticscrounger** [comment]
This would be less of a wish if we were able to specify whether we are looking for history items or bookmarks. It's main utility is to help navigate through large numbers of completion items. The command history is less important.

**Comment by The-Compiler** [comment]
Yeah, that's a lot more likely to happen, but unfortunately it'll probably take a few months until I get to it as it's not a priority right now.

**Comment by gnosticscrounger** [comment]
Fair enough. I'm trying to implement some of this functionality with keybindings. Would it be difficult to add an optional argument to the bind command that allows you to define a count? 

E.g, I want to define <Ctrl-E> in command mode to execute 6 counts of the completion-item-next command for quick scrolling of items.

**Comment by The-Compiler** [comment]
It's hard to say, there certainly are some corner-cases I can think of (what if you bind a command with a count and use the binding with a count?).

In your case, you could also use `:repeat` though:

`:bind --mode command <ctrl-e> repeat 6 completion-item-next`

**Comment by gnosticscrounger** [comment]
The above binds successfully but appears to do nothing.  Still, I did not know about the repeat command so thank you for sharing.

**Comment by The-Compiler** [comment]
That's odd, it works fine for me.

**Comment by gnosticscrounger** [comment]
I'll try deleting it from the config file and rebinding.

**Comment by gnosticscrounger** [comment]
Ok, just now I tried it and it started to magically work. Confused but happy!

**Comment by The-Compiler** [comment]
> Ok, just now I tried it and it started to magically work. Confused but happy!

The current config code sucks, you might just have run into some bug. Ripping it out and replacing it by something more sane is what I plan for after QtWebEngine support :wink:

I thought a bit more about this issue - while it's hard to do something like vim (i.e. have a buffer for those commands) it might be easier to render a webpage for the command history (or completion) and make them links, so you could use hints on them.

**Comment by gnosticscrounger** [comment]
That seems doable and imo more ideomatic for a web application.

**Comment by rcorre** [comment]
I was just thinking that `ctrl-n/ctrl-p` are a bit frustrating to use to find an old command.
I had a few ideas:

1. Bash-style `!`. `:!open` (or just `!open` if we wanted to add a new command prefix) would execute the last command starting with `open`. `!2` would execute the 2nd-previous command. Just entering `!` would pop a completion menu for previous commands.

2. `command-history-search`, a new command-mode-only command that works like readline's `ctrl+R` 

3. Add an optional argument to `repeat-command` to execute the n'th previous command, with completions searchable by index or text (similar to buffer completion)

**Comment by The-Compiler** [comment]
What about considering the already entered input for ctrl-p/ctrl-n, similar to how zsh (and bash?) can be configured? So with `:open <ctrl-p>` you'd only get commands starting with `:open`. Though I'm not sure how straightforward it'd be to implement, as of course the text filled via `<ctrl-p>` shouldn't be considered.

**Comment by lahwaacz** [comment]
If I type `:open` and press `<ctrl-p>`, I'm getting only commands starting with `:open` already :wink:

What about implementing similar functionality to `repeat-command`? For example `repeat-command open` would execute the last command starting with `open`. Then you could basically alias `!` to `repeat-command` (except for the whitespace between `repeat-command` and its first argument maybe).

**Comment by rcorre** [comment]
Hah, I just realized `ctrl-p` does that too :grin:

> What about implementing similar functionality to repeat-command

Yeah, that's what I was talking about in option 3. I think it would be the easiest and least invasive.

**Comment by The-Compiler** [comment]
Oh! :laughing: :blush: 

I wasn't sure if that was implemented, and then made a mistake when testing it.
